OpInfos for some Tensor dtype conversion methods

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #66358 OpInfos for: normal, bernoulli, multinomial
* #65941 OpInfo for `*_like` functions
* **#64282 OpInfos for some Tensor dtype conversion methods**

OpInfos for:
- Tensor.bfloat16, Tensor.bool, Tensor.bypte, Tensor.char
- Tensor.double, Tensor.float, Tensor.half, Tensor.int
- Tensor.short, Tensor.long

None of these are supported by TorchScript. Also, the OpInfo autograd
test runner assumes that the operation is not allowed to change the
dtype of the argument, so only Tensor.double has
`supports_autograd=True` (in theory Tensor.bfloat16, Tensor.float,
Tensor.half should be differentiable).

Test Plan:
- run tests

Differential Revision: [D31452627](https://our.internmc.facebook.com/intern/diff/D31452627)

zou3519 committed:
OpInfos for some Tensor dtype conversion methods

OpInfos for:
- Tensor.bfloat16, Tensor.bool, Tensor.bypte, Tensor.char
- Tensor.double, Tensor.float, Tensor.half, Tensor.int
- Tensor.short, Tensor.long

None of these are supported by TorchScript. Also, the OpInfo autograd
test runner assumes that the operation is not allowed to change the
dtype of the argument, so only Tensor.double has
`supports_autograd=True` (in theory Tensor.bfloat16, Tensor.float,
Tensor.half should be differentiable).

Test Plan:
- run tests

[ghstack-poisoned]

zou3519 committed:
Update on "OpInfos for some Tensor dtype conversion methods"


OpInfos for:
- Tensor.bfloat16, Tensor.bool, Tensor.bypte, Tensor.char
- Tensor.double, Tensor.float, Tensor.half, Tensor.int
- Tensor.short, Tensor.long

None of these are supported by TorchScript. Also, the OpInfo autograd
test runner assumes that the operation is not allowed to change the
dtype of the argument, so only Tensor.double has
`supports_autograd=True` (in theory Tensor.bfloat16, Tensor.float,
Tensor.half should be differentiable).

Test Plan:
- run tests

[ghstack-poisoned]

zou3519 committed:
Update on "OpInfos for some Tensor dtype conversion methods"


OpInfos for:
- Tensor.bfloat16, Tensor.bool, Tensor.bypte, Tensor.char
- Tensor.double, Tensor.float, Tensor.half, Tensor.int
- Tensor.short, Tensor.long

None of these are supported by TorchScript. Also, the OpInfo autograd
test runner assumes that the operation is not allowed to change the
dtype of the argument, so only Tensor.double has
`supports_autograd=True` (in theory Tensor.bfloat16, Tensor.float,
Tensor.half should be differentiable).

Test Plan:
- run tests

[ghstack-poisoned]

pmeier commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -5543,6 +5543,19 @@ def generator():
+
+    cases = ((),
+             (2, 3))
+    memory_format_options = [None, torch.preserve_format, torch.contiguous_format]
```
What about `torch.channels_last`?

```suggestion
    memory_format_options = [None, torch.preserve_format, torch.contiguous_format, torch.channels_last]
```

pmeier commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -5543,6 +5543,19 @@ def generator():
+
+    def generator():
+        for shape, memory_format in itertools.product(cases, memory_format_options):
+            yield(SampleInput(make_arg(shape, requires_grad=requires_grad),
```
Given that we unconditionally set `requires_grad` grad here, we can merge it into the `partial` above.

pmeier commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -5543,6 +5543,19 @@ def generator():
+def sample_inputs_conversion(op_info, device, dtype, requires_grad, **kwargs):
+    make_arg = partial(make_tensor, dtype=dtype, device=device)
+
+    cases = ((),
```
Maybe rename this to `shapes` to make it more readable

```suggestion
    shapes = ((),
```

pmeier commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -9221,6 +9234,99 @@ def generate_std_var_kwargs(t: torch.Tensor, **kwargs):
+           dtypes=all_types_and_complex_and(torch.bool, torch.half, torch.bfloat16),
+           supports_out=False,
+           sample_inputs_func=sample_inputs_conversion,
+           supports_autograd=False,
```
```suggestion
           # The autograd test runner cannot handle functions that change dtype
           supports_autograd=False,
```

pmeier commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -9221,6 +9234,99 @@ def generate_std_var_kwargs(t: torch.Tensor, **kwargs):
+           dtypes=all_types_and_complex_and(torch.bool, torch.half, torch.bfloat16),
+           supports_out=False,
+           sample_inputs_func=sample_inputs_conversion,
+           supports_autograd=False,
```
```suggestion
           # The autograd test runner cannot handle functions that change dtype
           supports_autograd=False,
```

pmeier commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -9221,6 +9234,99 @@ def generate_std_var_kwargs(t: torch.Tensor, **kwargs):
+           # The autograd test runner cannot handle functions that change dtype
+           supports_autograd=False,
+           skips=(
+               # This method not supported by TorchScript
+               DecorateInfo(unittest.skip("Skipped!"), 'TestJit', 'test_variant_consistency_jit'),
```
Given that this test is skipped by a lot of `OpInfo`'s we probably don't need a comment here. On the other hand, I would make the skip more precise by adding the skipped dtypes:
```suggestion
               DecorateInfo(unittest.skip("Skipped!"), 'TestJit', 'test_variant_consistency_jit', dtypes=(torch.float32, torch.complex64)),
```

This comment applies to all occurrences.

zou3519 replied:
Removed the comment. I didn't want to make the skip more precise: the reasoning is, `test_variant_consistency_jit` only seems to be run on some dtypes. However, no matter the dtype, `test_variant_consistency_jit` should fail the test because TorchScript simply doesn't support the operator. In the future, if someone comes along and adds more dtypes to `test_variant_consistency_jit`, then they would need to update the dtypes if we made it more precise.

pmeier reviewed (approved):
Only minor comments inline. Otherwise LGTM when CI is green!

zou3519 committed:
Update on "OpInfos for some Tensor dtype conversion methods"


OpInfos for:
- Tensor.bfloat16, Tensor.bool, Tensor.bypte, Tensor.char
- Tensor.double, Tensor.float, Tensor.half, Tensor.int
- Tensor.short, Tensor.long

None of these are supported by TorchScript. Also, the OpInfo autograd
test runner assumes that the operation is not allowed to change the
dtype of the argument, so only Tensor.double has
`supports_autograd=True` (in theory Tensor.bfloat16, Tensor.float,
Tensor.half should be differentiable).

Test Plan:
- run tests

[ghstack-poisoned]

zou3519 committed:
Update on "OpInfos for some Tensor dtype conversion methods"


OpInfos for:
- Tensor.bfloat16, Tensor.bool, Tensor.bypte, Tensor.char
- Tensor.double, Tensor.float, Tensor.half, Tensor.int
- Tensor.short, Tensor.long

None of these are supported by TorchScript. Also, the OpInfo autograd
test runner assumes that the operation is not allowed to change the
dtype of the argument, so only Tensor.double has
`supports_autograd=True` (in theory Tensor.bfloat16, Tensor.float,
Tensor.half should be differentiable).

Test Plan:
- run tests

[ghstack-poisoned]

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -9344,6 +9360,91 @@ def generate_std_var_kwargs(t: torch.Tensor, **kwargs):
+           # The autograd test runner cannot handle functions that change dtype
+           supports_autograd=False,
+           skips=(
+               DecorateInfo(unittest.skip("Skipped!"), 'TestJit', 'test_variant_consistency_jit'),
```
Comments for these skips would be nice, too

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -9344,6 +9360,91 @@ def generate_std_var_kwargs(t: torch.Tensor, **kwargs):
            dtypes=all_types_and_complex_and(torch.bool, torch.half, torch.bfloat16),
            sample_inputs_func=sample_inputs_scatter,),
+    OpInfo('Tensor.bfloat16',
+           dtypes=all_types_and_complex_and(torch.bool, torch.half, torch.bfloat16),
```
For ops that are method only, like contiguous, see here:

https://github.com/pytorch/pytorch/blob/6d4d636d66774eb803cffedd267849a8b8a3b536/torch/testing/_internal/common_methods_invocations.py#L6591

That's probably most convenient pattern to follow here, too?

mruberry reviewed (approved):
Approving for velocity but check inline review comments requesting a comment for the jit skips and suggesting modeling these after the current contiguous OpInfo which defines a functional variant using a lambda

zou3519 committed:
Update on "OpInfos for some Tensor dtype conversion methods"


OpInfos for:
- Tensor.bfloat16, Tensor.bool, Tensor.bypte, Tensor.char
- Tensor.double, Tensor.float, Tensor.half, Tensor.int
- Tensor.short, Tensor.long

None of these are supported by TorchScript. Also, the OpInfo autograd
test runner assumes that the operation is not allowed to change the
dtype of the argument, so only Tensor.double has
`supports_autograd=True` (in theory Tensor.bfloat16, Tensor.float,
Tensor.half should be differentiable).

Test Plan:
- run tests

[ghstack-poisoned]

zou3519 committed:
Update on "OpInfos for some Tensor dtype conversion methods"


OpInfos for:
- Tensor.bfloat16, Tensor.bool, Tensor.bypte, Tensor.char
- Tensor.double, Tensor.float, Tensor.half, Tensor.int
- Tensor.short, Tensor.long

None of these are supported by TorchScript. Also, the OpInfo autograd
test runner assumes that the operation is not allowed to change the
dtype of the argument, so only Tensor.double has
`supports_autograd=True` (in theory Tensor.bfloat16, Tensor.float,
Tensor.half should be differentiable).

Test Plan:
- run tests

Differential Revision: [D31452627](https://our.internmc.facebook.com/intern/diff/D31452627)

[ghstack-poisoned]

zou3519 committed:
Update on "OpInfos for some Tensor dtype conversion methods"


OpInfos for:
- Tensor.bfloat16, Tensor.bool, Tensor.bypte, Tensor.char
- Tensor.double, Tensor.float, Tensor.half, Tensor.int
- Tensor.short, Tensor.long

None of these are supported by TorchScript. Also, the OpInfo autograd
test runner assumes that the operation is not allowed to change the
dtype of the argument, so only Tensor.double has
`supports_autograd=True` (in theory Tensor.bfloat16, Tensor.float,
Tensor.half should be differentiable).

Test Plan:
- run tests

Differential Revision: [D31452627](https://our.internmc.facebook.com/intern/diff/D31452627)

[ghstack-poisoned]

zou3519 committed:
Update on "OpInfos for some Tensor dtype conversion methods"


OpInfos for:
- Tensor.bfloat16, Tensor.bool, Tensor.bypte, Tensor.char
- Tensor.double, Tensor.float, Tensor.half, Tensor.int
- Tensor.short, Tensor.long

None of these are supported by TorchScript. Also, the OpInfo autograd
test runner assumes that the operation is not allowed to change the
dtype of the argument, so only Tensor.double has
`supports_autograd=True` (in theory Tensor.bfloat16, Tensor.float,
Tensor.half should be differentiable).

Test Plan:
- run tests

Differential Revision: [D31452627](https://our.internmc.facebook.com/intern/diff/D31452627)

[ghstack-poisoned]

zou3519 committed:
Update on "OpInfos for some Tensor dtype conversion methods"


OpInfos for:
- Tensor.bfloat16, Tensor.bool, Tensor.bypte, Tensor.char
- Tensor.double, Tensor.float, Tensor.half, Tensor.int
- Tensor.short, Tensor.long

None of these are supported by TorchScript. Also, the OpInfo autograd
test runner assumes that the operation is not allowed to change the
dtype of the argument, so only Tensor.double has
`supports_autograd=True` (in theory Tensor.bfloat16, Tensor.float,
Tensor.half should be differentiable).

Test Plan:
- run tests

Differential Revision: [D31452627](https://our.internmc.facebook.com/intern/diff/D31452627)

[ghstack-poisoned]

zou3519 commented:
@zou3519 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31452627).

zou3519 committed:
Update on "OpInfos for some Tensor dtype conversion methods"


OpInfos for:
- Tensor.bfloat16, Tensor.bool, Tensor.bypte, Tensor.char
- Tensor.double, Tensor.float, Tensor.half, Tensor.int
- Tensor.short, Tensor.long

None of these are supported by TorchScript. Also, the OpInfo autograd
test runner assumes that the operation is not allowed to change the
dtype of the argument, so only Tensor.double has
`supports_autograd=True` (in theory Tensor.bfloat16, Tensor.float,
Tensor.half should be differentiable).

Test Plan:
- run tests

Differential Revision: [D31452627](https://our.internmc.facebook.com/intern/diff/D31452627)

[ghstack-poisoned]
